File: app/SourcesShared/RatingsFormat.swift
import Foundation

/// ONE consistent formatting for cross-provider ratings across every VortX surface: the landscape card
/// badge (`CardRatingBadge`), the detail hero's primary rating position (`iOSDetailView` / tvOS
/// `DetailView`), and the server-side baked poster (mirrored in the `cloudflare-poster` worker's
/// `buildSvg`). Change the labels/order/number format here and every Apple surface moves together, so the
/// card, the detail row, and the baked poster never disagree about how a score reads.
///
/// The numbers cannot drift because every surface reaches them the same way: `VortXRatings.swift` and
/// `MDBListClient.swift` both land the aggregators with `Int($0.rounded())`, which is exactly what the
/// worker's `Math.round` does, and both read the same service keys (imdb / rt / metacritic / tmdb). The
/// worker's own unit test asserts the printed runs in this order, so a change on that side is caught; a
/// change on THIS side has no test behind it, so anything edited here must be mirrored by hand.
///
/// What DOES differ, deliberately, is the arrangement. The card draws these tokens as the JEWEL capsule
/// (a glass plate, lettermark beside its value) because it floats in a corner over unpredictable artwork;
/// the baked portrait poster draws them as the MARQUEE block (type straight on the art, lettermark stacked
/// above its value under a rule) because it owns a tall canvas with room for a real composition. That is a
/// design decision, not drift: do not "reunify" the two compositions. Only the tokens are shared.
///
/// The model is `MDBListRatings` (the shared decode target for both the keyless VortX ratings service and a
/// user's own MDBList key): IMDb on its native 0-10 scale, Rotten Tomatoes / TMDB as 0-100 percentages, and
/// Metacritic as a 0-100 metascore (no percent sign, matching how the score is conventionally printed).
enum RatingsFormat {
    /// A single provider score as a label + value, e.g. ("RT", "84%"). `isIMDb` marks the primary score that
    /// leads with a star glyph on the badges (its label is dropped there in favor of the star), while the
    /// detail row prints the label verbatim.
    struct Token: Equatable {
        let label: String
        let value: String
        let isIMDb: Bool
    }

    /// IMDb on its native 0-10 scale, one decimal ("8.3"). Matches the baked poster's `imdb.toFixed(1)`, so
    /// the card, the detail row, and the poster all print the IMDb number identically.
    static func imdb(_ v: Double) -> String { String(format: "%.1f", v) }

    /// Ordered tokens for every score present, IMDb -> RT -> MC -> TMDB. Empty when nothing is present.
    /// `limit` caps how many tokens a size-constrained surface renders (nil = all four): a small card asks
    /// for two (star + RT), a tvOS card or the detail row for all of them. Per-score fail-soft by
    /// construction: a title carrying only IMDb yields exactly one token, so it renders only IMDb.
    static func tokens(_ r: MDBListRatings, limit: Int? = nil) -> [Token] {
        var out: [Token] = []
        if let v = r.imdb { out.append(Token(label: "IMDb", value: imdb(v), isIMDb: true)) }
        if let v = r.rottenTomatoes { out.append(Token(label: "RT", value: "\(v)%", isIMDb: false)) }
        if let v = r.metacritic { out.append(Token(label: "MC", value: "\(v)", isIMDb: false)) }
        if let v = r.tmdb { out.append(Token(label: "TMDB", value: "\(v)%", isIMDb: false)) }
        if let limit, limit >= 0, out.count > limit { return Array(out.prefix(limit)) }
        return out
    }

    /// The joined detail-row string ("IMDb 8.3  ·  RT 84%  ·  MC 69  ·  TMDB 81%"), or nil when no score is
    /// present. The one string every detail surface prints, so tvOS and iOS/Mac never diverge.
    static func joined(_ r: MDBListRatings) -> String? {
        let parts = tokens(r).map { "\($0.label) \($0.value)" }
        return parts.isEmpty ? nil : parts.joined(separator: "  \u{00B7}  ")
    }
}

File: android/app/src/main/kotlin/com/vortx/android/ratings/RatingsFormat.kt
package com.vortx.android.ratings

import java.util.Locale

/// ONE consistent formatting for cross-provider ratings across the Android detail surface, the Kotlin port
/// of Apple `app/SourcesShared/RatingsFormat.swift`. IMDb leads with an accent star (its label dropped), then
/// Rotten Tomatoes, then Metacritic, then TMDB, formatted once here so the hero token strip and any future
/// card badge never disagree about how a score reads. The model is [MdbListRatings] (IMDb on its native 0-10
/// scale, RT / TMDB as 0-100 percentages, Metacritic as a bare 0-100 metascore), so the same shape drives
/// both the keyless VortX ratings service and a user's own MDBList key.
object RatingsFormat {

    /// A single provider score as a label + value, e.g. ("RT", "84%"). [isImdb] marks the primary score that
    /// leads with the accent star (its label is dropped in favour of the star); every other token prints its
    /// label verbatim. Mirrors Apple `RatingsFormat.Token`.
    data class Token(val label: String, val value: String, val isImdb: Boolean)

    /// IMDb on its native 0-10 scale, one decimal ("8.3"). Locale-fixed (US) so the decimal separator matches
    /// the baked poster's `toFixed(1)` regardless of device locale. Mirrors Apple `RatingsFormat.imdb`.
    fun imdb(value: Double): String = String.format(Locale.US, "%.1f", value)

    /// Ordered tokens for every score present, IMDb -> RT -> MC -> TMDB. Empty when nothing is present.
    /// [limit] caps how many tokens a size-constrained surface renders (null = all four). Per-score
    /// fail-soft by construction: a title carrying only IMDb yields exactly one token. Mirrors Apple
    /// `RatingsFormat.tokens`.
    fun tokens(r: MdbListRatings, limit: Int? = null): List<Token> {
        val out = ArrayList<Token>(4)
        r.imdb?.let { out += Token("IMDb", imdb(it), isImdb = true) }
        r.rottenTomatoes?.let { out += Token("RT", "$it%", isImdb = false) }
        r.metacritic?.let { out += Token("MC", "$it", isImdb = false) }
        r.tmdb?.let { out += Token("TMDB", "$it%", isImdb = false) }
        if (limit != null && limit >= 0 && out.size > limit) return out.take(limit)
        return out
    }

    /// The joined detail-row string ("IMDb 8.3  ·  RT 84%  ·  MC 69  ·  TMDB 81%"), or null when no score is
    /// present. Mirrors Apple `RatingsFormat.joined`.
    fun joined(r: MdbListRatings): String? {
        val parts = tokens(r).map { "${it.label} ${it.value}" }
        return if (parts.isEmpty()) null else parts.joinToString("  ·  ")
    }
}
